### PR TITLE
F/netmask6 fixes

### LINKED ISSUE #475: Add a warning message to netmask6 filter driver when filter is invalid.

### LINKED ISSUE #476: Fix netmask6 filter on bigendian systems

### PR DESCRIPTION
**Pull request by deirf**
fixes #476 
fixes #475

### CODE REVIEW

**Comment by dnsjts** [comment]
@deirf: I have submitted a pull request to your repository, please merge/reject it before any rebase.

**Comment by ihrwein** [comment]
@deirf : Is this PR supposed to solve https://github.com/balabit/syslog-ng/issues/476? If that's so please add `fixes #476` to the description of your PR.

**Comment by gregory094** [comment]
done

File: lib/filter/filter-expr-grammar.ym [hunk 1]
```diff
@@ -103,12 +103,13 @@ filter_simple_expr
...
-                                    $$ = filter_netmask6_new($3);
+    if (!$3) YYERROR;
+    $$ = filter_netmask6_new($3);
   #else
-                                    YYERROR;
+    YYERROR;
   #endif
-                                    free($3);
-                                  }
+    free($3);
```
Comment (ihrwein) [inline lib/filter/filter-expr-grammar.ym]: Is the `free()` call in the good place? I think we should put it after `filter_netmask6_new()` and before the 108th line.

File: lib/filter/filter-netmask6.c [hunk 2]
```diff
@@ -114,42 +110,80 @@ _eval(FilterExprNode *s, LogMessage **msgs, gint num_msg)
...
+  filter_netmask6_calc_network_address((unsigned char *) &address, self->prefix, &network_address);
+  result = _in6_addr_equals(&network_address, &self->address);
 
   return result ^ s->comp;
 }
 
-FilterExprNode *
-filter_netmask6_new(gchar *cidr)
+static void
+_get_validated_address(FilterNetmask6 *self, gchar *cidr, gchar address[])
```
Comment (gregory094) [inline lib/filter/filter-netmask6.c]: This should be splitted into two different functions for it has multiple responsibilities. 
* for normalizing the key: `normalize_address`
* for the settings the defaults if address cannot be normalized: `set_address_defaults()`

**Comment by deirf** [comment]
@lbudai could you look at this PR?
I know there is a "makefile and ifdef" review note, I will do it and send new PR about it.

**Comment by lbudai** [comment]
As I know, Gergo fixed the makefile related issues, so talk to him and when you think, you are finished (and don't have any unprocessed review note), just ping me.

I have lots of todos, and I've a priority list on patches.
Bazsi sent some PRs and I want to start with those PRs.

**Comment by gregory094** [comment]
You can check my solution [here](https://github.com/gregory094/syslog-ng/tree/f/netmask6).

File: lib/filter/filter-netmask6.c [hunk 3]
```diff
@@ -114,42 +109,105 @@ _eval(FilterExprNode *s, LogMessage **msgs, gint num_msg)
...
+  gchar *slash = strchr(cidr, '/');
+
+  if (slash)
+    return slash - cidr;
+  else
+    return INET6_ADDRSTRLEN;
+}
+
+static gboolean
+_get_noramlized_address(FilterNetmask6 *self, gchar *cidr, gchar address[])
```
Comment (gregory094) [inline lib/filter/filter-netmask6.c]: 'get_normalized_address'

File: lib/filter/filter-netmask6.c [hunk 4]
```diff
@@ -114,42 +109,105 @@ _eval(FilterExprNode *s, LogMessage **msgs, gint num_msg)
...
+      msg_warning("Mask is empty; setting to default (128)",
+                  NULL);
+      return default_prefix;
+    }
+
+  return atol(slash + 1);
+}
+
+static gint
+_get_noramlized_address_len(gchar *cidr)
```
Comment (gregory094) [inline lib/filter/filter-netmask6.c]: 'get_normalized_address'

File: lib/filter/filter-netmask6.c [hunk 5]
```diff
@@ -114,42 +109,105 @@ _eval(FilterExprNode *s, LogMessage **msgs, gint num_msg)
...
-  get_network_address((unsigned char *) &address, self->prefix, &network_address);
-  result = _in6_addr_compare(&network_address, &self->address);
+  filter_netmask6_calc_network_address((unsigned char *) &address, self->prefix, &network_address);
+  result = _in6_addr_equals(&network_address, &self->address);
 
   return result ^ s->comp;
 }
 
+static gint
+_get_noramlized_prefix(FilterNetmask6 *self, gchar *cidr)
```
Comment (gregory094) [inline lib/filter/filter-netmask6.c]: 'get_normalized_address'

File: lib/filter/filter-netmask6.c [hunk 6]
```diff
@@ -114,42 +109,105 @@ _eval(FilterExprNode *s, LogMessage **msgs, gint num_msg)
...
+  gchar *slash = strchr(cidr, '/');
+
+  if (!slash)
+    {
+      msg_warning("Mask is empty; setting to default (128)",
+                  NULL);
+      return default_prefix;
+    }
+
+  return atol(slash + 1);
```
Comment (gregory094) [inline lib/filter/filter-netmask6.c]: what about 'strtol' ?

File: lib/filter/filter-netmask6.c [hunk 7]
```diff
@@ -114,42 +108,123 @@ _eval(FilterExprNode *s, LogMessage **msgs, gint num_msg)
...
+  self->prefix = prefix;
+
+  if (inet_pton(AF_INET6, address, &packet_addr) == 1)
+    filter_netmask6_calc_network_address((unsigned char *) &packet_addr, self->prefix, &self->address);
   else
-    self->address = in6addr_loopback;
+    goto filter_netmask6_invalid_error;
 
   self->super.eval = _eval;
   return &self->super;
```
Comment (gregory094) [inline lib/filter/filter-netmask6.c]: Possibly it's a stupid question but we are leaking 'self' here... only `FilterExprNode` will be freed, `FilterNetmask6` isn't. Is it a valid point? @deirf

Comment (deirf) [inline lib/filter/filter-netmask6.c]: Self will be freed in filter_expr_unref() function.

**Comment by deirf** [comment]
@ gregory094 I saw you solution please check my modifications and my answers your proposals.

File: lib/filter/filter-expr-grammar.ym [hunk 8]
```diff
@@ -103,12 +105,11 @@ filter_simple_expr
 	| KW_NETMASK '(' string ')'    { $$ = filter_netmask_new($3); free($3); }
   | KW_NETMASK6 '(' string ')'   {
   #if ENABLE_IPV6
-                                    $$ = filter_netmask6_new($3);
-  #else
-                                    YYERROR;
+    $$ = filter_netmask6_new($3);
```
Comment (ihrwein) [inline lib/filter/filter-expr-grammar.ym]: What happens when syslog-ng is configured without IPv6 support but the user tries to use this filter?

File: lib/filter/filter-netmask6.c [hunk 9]
```diff
@@ -114,42 +115,151 @@ _eval(FilterExprNode *s, LogMessage **msgs, gint num_msg)
...
+  return FALSE;
+}
+
+static inline gboolean
+_is_valid_address_len(gint len)
+{
+  return (len > 0) && (len <= INET6_ADDRSTRLEN);
+}
+
+static inline  gboolean
```
Comment (ihrwein) [inline lib/filter/filter-netmask6.c]: Please remove the extra space before `gboolean`.

File: lib/filter/filter-expr-grammar.ym (hunk 8 above)
Comment (deirf) [inline lib/filter/filter-expr-grammar.ym]: Error parsing filter expression, filter plugin netmask6 not found OR you may not used double quotes in your filter expression in....

**Comment by kira-syslogng** [comment]
Can one of the admins verify this patch?

**Pull request closed without merging**